packages/@tailwindcss-upgrade/src/index.ts

```diff
@@ -151,42 +151,23 @@ async function run() {
 
     // Cleanup `@import "…" layer(utilities)`
     for (let sheet of stylesheets) {
-      // If the `@import` contains an injected `layer(…)` we need to remove it
-      if (!Array.from(sheet.importRules).some((node) => node.raws.tailwind_injected_layer)) {
-        continue
-      }
-
-      let hasAtUtility = false
-
-      // Only remove the `layer(…)` next to the import, if any of the children
-      // contains an `@utility`. Otherwise the `@utility` will not be top-level.
-      {
-        sheet.root.walkAtRules('utility', () => {
-          hasAtUtility = true
-          return false
-        })
-
-        if (!hasAtUtility) {
-          for (let child of sheet.descendants()) {
-            child.root.walkAtRules('utility', () => {
-              hasAtUtility = true
-              return false
-            })
-
-            if (hasAtUtility) {
-              break
-            }
-          }
+      for (let importRule of sheet.importRules) {
+        if (!importRule.raws.tailwind_injected_layer) continue
+        let importedSheet = stylesheets.find(
+          (sheet) => sheet.id === importRule.raws.tailwind_destination_sheet_id,
+        )
+        if (!importedSheet) continue
+
+        // Only remove the `layer(…)` next to the import, if any of the children
+        // contains an `@utility`. Otherwise the `@utility` will not be top-level.
```

Comment:
```suggestion
        // Only remove the `layer(…)` next to the import if any of the children
        // contain `@utility`. Otherwise `@utility` will not be top-level.
```